### ISSUE TITLE
Add --gencode option to Alignment-based mode

### ISSUE THREAD

**Issue by JohnMCMa** [COMBINE-lab/salmon#95]
Currently the indexer has a --gencode flag to handle the composite fasta headers in GENCODE transcript fasta files and `split the transcript name at the first '|' character`.

However, the alignment-based mode doesn't use an index, and the current behavior is to read the fasta headers from -t verbatim. So, if I run the STAR aligner with the standard workflow using GENCODE files as index, the Aligned.toTranscriptome.out.bam output would already be aligned to the transcriptome, with the RNAME fields set in the same way `salmon index --gencode`, for example:

`ST-J00106:110:H5NY5BBXX:3:2211:22495:7240       99      ENSMUST00000195335.1    799     60      76M     =       945     222     CTCAGTAGGAAGATTATAACTAATACTCCCCCATCAAACAGTTTTAAGGACAGAAGAGAACAAAGCATGTAAAGTG  <AFFFJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJJFJJJJJJJJJJJJJJJJFFJA<AFJJFFA NH:i:1   HI:i:1  RG:Z:160212_ST-J00106_0110.3`

As a result, Salmon won't quantify anything if I run Salmon in the Alignment-based mode with the official GENCODE transcripts.fa, for example, `salmon quant -l IU -a Aligned.toTranscriptome.out.bam -o $OUTPUT_DIR -t gencode.vM9.pc_transcripts.fa -g gencode.vM9.annotation.gtf` since the RNAME and the FASTA headers don't match, causing a large number of messages like `WARNING: Transcript ENSMUST00000082421.1|ENSMUSG00000064370.1|-|-|mt-Cytb-201|mt-Cytb|1144|CDS:1-1144| appears in the reference but did not appear in the BAM`

Similarly, the gene quantification mode won't work either, since the names in -t and -g will never match.

While for the immediate need I can just trim the fasta headers from the GENCODE transctipts.fa using a simple script, it must be an oversight that the same functionality is provided in quasi-alignment mode only.

**Comment by rob-p** [comment]
Hi John,

  Thanks for reporting this.  Indeed this is an oversight.  The gencode option is applied in quasi-mapping mode during index construction,  but there is no indexing step in alignment mode.  So, this transformation should be applied when the reference sequences are loaded in alignment-based mode.  I'll add this option to alignment-based quant for v0.7.3, which should be coming soon.  Thanks again for reporting this.

Best,
Rob

**Comment by JohnMCMa** [comment]
Sorry for responding late, since I haven't been dealing with RNA-seq data lately. Fix confirmed to work as of version 0.8.2; closed.

**Comment by callumparr** [comment]
Sorry to add on previous issue but I am trying to run salmon quant using --gencode flag but it still outputs errors that header names in BAM file cannot be found in the input fasta. 

gencode.M24.transcript.fa was indexed by minimap2 and reads were aligned using map-ont mode, using the unsorted BAM file ran the following command: 

```
salmon quant --noErrorModel --gencode -p 4 -t /home/callum/Genome_files/gencode.vM24.transcripts.fa -l U -a alignments/O1.bam -o counts/O1_salmon
```

Example line from the error output. 
```
[2021-06-04 13:48:14.728] [jointLog] [critical] Transcript ENSMUST00000078393.6|ENSMUSG00000062279.6|OTTMUSG00000071892.1|OTTMUST00000173863.1|Gm10237-201|Gm10237|336|processed_pseudogene| appeared in the BAM header, but was not in the provided FASTA file 
```
